Add RTP demux algorithm

pthatcherg opened the pull request.

fluffy commented:
I think this is close enough to merge it, put it out in a draft, and see what the WG things

ekr commented:
This LGTM.

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+        <section title="Processing RTP packets"
+                 anchor="sec.rtp.demux">
+          <t>When an RTP packet is received by a transport (typically a
+          DtlsTransport), that packet MUST be routed to the correct
```
The term 'DtlsTransport' has not been defined in this document. Suggest omitting the parenthesized text.

ekr replied:
This first MUST seems less normative and more like "this is what has to happen". Perhaps "needs to"

pthatcherg replied:
Removed DtlsTransport and the first MUST

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+          <t>
+            <list>
+              <t>Construct a table mapping MID to RtpReceiver. For each
+              RtpReceiver configured to recei ve from this transport,
```
typo: space

pthatcherg replied:
Fixed

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              let the key be the RtpReceiver's MID, and the value be the
+              RtpReceiver.</t>
+              
+              <t>Construct a table mapping SSRC to RtpReceiver. For each
```
How? From what data?

pthatcherg replied:
transceiver.receiver.getParameters().encodings[j].ssrc, transceiver.receiver.getParameters().encodings[j].rtx.ssrc, and transceiver.receiver.getParameters().encodings[j].fec.ssrc

are all defined in the W3C doc, and those values are ultimately filled in by the remote description, which is defined in this document.  So shouldn't we be able to just say "via the SSRCs found in the remote description"?

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+
+          <t>
+            <list>
+              <t>Construct a table mapping MID to RtpReceiver. For each
```
Where do the MIDs come from? (I know, but think we should spell it out)

pthatcherg replied:
Same here: transceiver.mid is defined in the W3C doc, but that's just a reflection of what's in the offer description (local or remote), so it seems like we ought to be able to point back to this document for those.

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              for each SSRC that RtpReceiver is configured to receive,
+              let the key be the SSRC and the value be the RtpReceiver.</t>
+
+              <t>Construct a table mapping Payload Type to RtpReceiver. For
```
again, how, etc

pthatcherg replied:
Same here: the PTs in the W3C spec are read-only reflections of the PTs in the local description, so we should be able to reference the local description instead.

pthatcherg replied:
RIDs are a different level of processing.  RIDs are used to by a given RtpReceiver to know how to process a set of incoming RTP streams.  But this algorithm is for deciding which RtpReceiver to send an RTP stream to.  If RIDs were used without MIDs, I suppose you could use RIDs for that, but RIDs are intended to be used with MIDs and be scoped at the MID level, so I think I'd prefer using MIDs for this and leave RIDs to be used by the RtpReceiver after its been chosen by this algorithm.

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              <t>Construct a table mapping MID to RtpReceiver. For each
+              RtpReceiver configured to recei ve from this transport,
+              let the key be the RtpReceiver's MID, and the value be the
+              RtpReceiver.</t>
```
also - need to make it clear that this is only for RtpReceivers associated with this transport.

pthatcherg replied:
The first paragraph says "For each transport, the following steps MUST be followed to prepare to route packets" and then each subsequent paragraph says "For each RtpReceiver configured to receive from this transport".  Does that not make it clear?

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+                the RtpReceiver.
+              </t>
+              <t>
+                If the packet's SSRC is in the SSRC mapping table,
```
Should this check come first (i.e. if we have the SSRC in the table, do we even need to check mid?)

ekr replied:
I have no strong opinion on the order of checks. I guess the semantic question is: say you get a packet with a known SSRC but a bogus MID, what behavior do we want?

aboba replied:
If the MID entry is bogus, the packet needs to be dropped, because otherwise it could be delivered to an inappropriate receiver.  Imagine that a MID was valid at one point so the SSRC has been "latched" (entered into the SSRC table).  Now the MID is no longer in the MID table (e.g. transceiver.stop() has been called and negotiation has concluded) but the "latched" SSRC entry is still there and due to delays there are still RTP packets with the now invalid MID in transit. You would not want to attempt to deliver those packets to transceiver.receiver.

pthatcherg replied:
The first rule drops a packet with a bogus mid, so that's already accounted for.  The second and third rules (then ones we are discussing) only matter in the case where you either don't have a MID (in which case the second rule doesn't apply) or you have a known MID (in which case the SSRC will be latched).

The edge case where the order matters is if the MID of an SSRC changes.  In that case "MID first" allows the packet to be routed to the new MID, but "SSRC first" routes it to the old MID.

It's a weird case, but I think routing to the new MID makes more sense.

fluffy replied:
I think what peter has in the the PR is right the right thing here in that a MID needs to override a previously latch SSRC. This can come up when SSRC are changing to deal with SSRC conflicts and stuff with the same MID has been changed to a new SSRC so that some other MID can use the old SSRC.

fluffy replied:
Big discussion in room down here and Ekr, Justin, Martin and I think you have this in the right order no change needed

ekr replied:
Agreed.

juberti commented:
overall, this looks like it's on the right track. Main question (for @ekr really) is whether this should be turned more into prose rather than a W3C-style algorithm.

(i.e. "First, check this. Then, do that. Next, do this other thing. Finally, drop the packet.")

fluffy commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
         </section>
+        <section title="Processing RTP packets"
+                 anchor="sec.rtp.demux">
+          <t>When an RTP packet is received by a transport (typically a
```
I think right here we should put something like "Note: The algorithm described in this section does not yet have WG consensus but is included here so there is something concrete for the WG to decide if they are OK with." so that we don't get howls from some people.

pthatcherg replied:
Note added.

fluffy commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             from any other implementation of this specification.</t>
         </section>
+        <section title="Processing RTP packets"
+                 anchor="sec.rtp.demux">
```
It might help to have a bit of intro text here that helps people realize that a Receiver corresponds to a m-line and then there is a second stage of demux that gets it to the correct codec on the m-line.  I'm thinking of something along the lines of 

When using RTP without bundle, each m-line had a separate port that can be use to know which m-line an incoming RTP packets was associated with. With bundle there are additional identifiers that are used in getting the RTP packet to the correct m-line. This section describes how the information found in the SDP and the RTP packet are used to associated the packet with the correct m-line. Once this has happened, the same approaches and non bundle RTP are used to get the packet to the correct codec in that m-line. A RtpReceiver will correspond to a exactly one m-line in the SDP. An RtpReceiver can have multiple codecs associated with it each with their own RID and SSRC. The RID values are allocated by the receiver and send to the server in the SDP while the SSRC are allocated by the sender. Using RID to demultiplex allows the receiver to demultiplex incoming RTP using the RID values before the SDP answer is received.

fluffy replied:
I think EKR would like some general theory of operation note that gets people the big idea of what algo is going to do. I am thinking something along lines of 

The main idea is when a RID is present, it overrides any SSRC information in deciding how to route the packet but updates the SSRC map such that future packets on the same SSRC but no RID get routed the same way. If no RIDs are ever received, and SSRC signaling information is known, then that can be used but RID information overrides SSRC if they happen to conflict as some architectures  may have devices that change the SSRC but nothing should be changing the RIDs.

pthatcherg replied:
I don't think we should demux *between* RtpReceivers using RID.  RIDs are scoped to MIDs, and so you would need to be careful to make all your RIDs unique across and entire transport for that to work.  It's much more safe to just use MIDs for demuxing between RtpReceivers, and then only use RIDs within an RtpReceiver to known which RTP streams repair which other RTP streams.

You can demux before receiving an answer using MIDs first (to choose between the RtpReceivers) and RIDs second (to know which RTP stream a repair stream repairs).

I added a paragraph at the end to try and make this more clear.

ekr replied:
@fluffy did you mean MID above

fluffy commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+                 anchor="sec.rtp.demux">
+          <t>When an RTP packet is received by a transport (typically a
+          DtlsTransport), that packet MUST be routed to the correct
+          RtpReceiver. For each transport, the following steps MUST be
```
Do we need to make it just a bit more explicit that if there are 2 UDP ports in use, then there are two transports and separate sets of tables for both ? What you have now is totally correct, I just worry that people reading this for first time might miss this is all per transport.  Don't feel strongly about this, if you think it is fine as it, works for me

ekr replied:
It's unusual to specify a specific algorithm. Are you actually requiring this or just the same outcomes or something else? This gets to the topic below about the priority of various indicia

ekr replied:
I think this needs some introductory text explaining what we are trying to achieve.

aboba replied:
I also believe that we need to describe how table entries are brought into existence or removed (and what methods are involved).  Since the table entries are per DtlsTransport, they cannot exist prior to construction of the DtlsTransport (e.g. when receiver.transport is null).  

IMHO, for an Offerer, table entries are set/removed by calling setRemoteDescription, and for an Answerer, by calling setLocalDescription.  

Since transceiver.setDirection and transceiver.stop set the negotiation-needed flag, I do not believe that these methods result in the immediate removal of table entries  (e.g. calling transceiver.setDirection(direction) with a direction value of "sendonly" or "inactive" does not remove table entries until negotiation completes).

ekr commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,65 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             process, as a remote offer or answer, unmodified SDP coming
             from any other implementation of this specification.</t>
         </section>
+        <section title="Processing RTP packets"
```
This text needs to clearly say that the SRTP has been validated before this algorithm is executed, because it's otherwise not safe to update the SSRC table based on anything you see in packets.

pthatcherg replied:
Added

fluffy commented:
Peter, if there is anything I can do to help on this one today. Ping me and I will make it top priority.

pthatcherg commented:
A response to Bernard's comment about when the tables are constructed:

I think it's clear that the tables need to be constructed after the transport exists, but before the first packet is routed.  But an implementation could choose whether to construct it right after the transport is created or right before the first packet is routed.  I don't think we need to specify that.  

You do bring up a really good point about updating the table over time.  We know entries can be added by setLocalDescription/setRemoteDescription and removed by setLocalDescription/setRemoteDescription and RtpTransceiver.stop.   But that's not covered at all by the text written so far.

fluffy commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,84 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+                entry mapping the packet's SSRC to the
+                RtpReceiver. Deliver the packet to the RtpReceiver and
+                stop.
+              </t>
```
I suspect that this step needs to be above the use SSRC in previous step. Imagine the case where all the PT were unique yet we had a SSRC conflict.

pthatcherg replied:
If PTs are *not* unique, then putting step above the SSRC rule would break the case where SSRCs are signaled, and the case where MIDs are used but then SSRCs are latched.

Can you explain more about SSRC conflicts and how would that disrupt demux?  Maybe we need one of those unit tests Bernard was talking about.

fluffy replied:
If the PT's conflict with the SSRCs then I think it fine to drop all the packets or let PT override because clearly that should not be happening. In the case where SSRC were signalling, but then something changed them, it still seems like PT should override SSRC. I think the place you want to use SSRC when there is no MID is just to disambiguate  between multiple things that have the PT.

pthatcherg replied:
Sorry, I'm not getting it.  Could you provide a more explicit example of the configuration and packets received where the current algorithm doesn't get the desired output?  In other words, if we had a mechanism for running the algorithm against test cases, what does the test case for this look like?

fluffy replied:
after much hashing on whiteboard we have decided you are brilliant and this is correct as you have it.

fluffy commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,84 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              Payload Type and the value be the RtpReceiver.  The
+              Payload Types of a given RtpReceiver are found in the
+              a=fmtp lines of the m= section corresponding to that
+              RtpReceiver in the local description.</t>
```
Given we can have duplicate PT, I think we need to be clear which one ends up in the table. One way  would be saying that they are added to the table in the order the "a=rtpmap" occur in the SDP, and that a PT is only added to the table if there is no mapping yet for the PT.

pthatcherg replied:
Perhaps the easiest thing to do is say "if there are duplicate RtpReceivers for the same key, use the  RtpReceiver whose m= section appears first in the SDP".   Such a rule would apply also to the table as it is updated over time, not just when it is constructed (which we still need to resolve)

fluffy replied:
works for me

pthatcherg replied:
Added, although I made the text say "If any Payload Type could map to more than one RtpReceiver, map to the RtpReceiver whose m= section appears earliest in the local description"

fluffy commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,84 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+          <t>
+            After packets are routed to the RtpReceiver, further
+            processing of the RTP packets is done at the RtpReceiver
+            level.  This includes using the RID to determine which RTP
```
Perhaps add ref to draft-ietf-mmusic-rid

pthatcherg replied:
Added

ekr commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4078,28 +4078,41 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
-              RtpReceiver.</t>
+              <t>Construct a table mapping MID to RtpReceiver. For
+              each RtpReceiver configured to receive from this
+              transport, let the key be the RtpReceiver's MID, and the
```
This "let the key be" stuff is duplicative of a table mapping A to B.

pthatcherg replied:
Removed the duplicate stuff

pthatcherg commented:
FYI, I updated this to address Bernard's comment about updating the tables over time.  I added a paragraph saying what the results of updates should be, but leaves the when and how up to the implementation.

martinthomson commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,95 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              section corresponding to that RtpReceiver in the local
+              description.  If any Payload Type could map to more than
+              one RtpReceiver, map to the RtpReceiver whose m= section
+              appears earliest in the local description.</t>
```
At this level, aren't we more interested in the payload type numbers rather than the codecs themselves?

pthatcherg replied:
Yes.  When I say "Payload Type" in a document, does that mean the number or the codec?  If the codec, what's the right term for the number?

juberti replied:
number = payload type
codec = media format

martinthomson commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,95 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+          <t>
+            <list>
+              <t>
+                If the packet has a MID and that MID is not in the table mapping MID to RtpReceiver, drop the packet and stop.
```
Nit: wrap

pthatcherg replied:
Fixed

martinthomson commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,95 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+            so in a way such that whenever the tables are used for
+            routing, they are equivalent to rebuilding the tables from
+            scratch and re-applying all updates to the SSRC mapping
+            table that were made by packets previously received.
```
I think that the last sentence here isn't necessary.

pthatcherg replied:
Great.  I removed it.

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,92 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
         </section>
+        <section title="Processing RTP packets"
+                 anchor="sec.rtp.demux">
+          <t>Note: The following algorithm does not yet have WG
```
Remove trailing quote

pthatcherg replied:
Done

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,92 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+          <t>Note: The following algorithm does not yet have WG
+          consensus but is included here as something concrete for the
+          working group to discuss."</t>
+          <t>When an RTP packet is received by a transport and the
```
"and the SRTP has been validated" sounds odd. 

Perhaps
"and passes SRTP authentication"

pthatcherg replied:
Done

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,92 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+            <list>
+              <t>Construct a table mapping MID to RtpReceiver for each
+              RtpReceiver configured to receive from this transport.
+              The MID of a given RtpReceiver is found in the a=mid
```
I don't think you need to say where MID comes from since it is discussed elsewhere. It's more complicated than this.

pthatcherg replied:
Removed

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,92 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              <t>Construct a table mapping SSRC to RtpReceiver for
+              each RtpReceiver configured to receive from this
+              transport and for each SSRC that RtpReceiver is
+              configured to receive.  The SSRCs of a given RtpReceiver
```
That's not the only place where the SSRCs can come from, and they're not required. I would say to populate the table using the supplied SSRCs, if present.

pthatcherg replied:
Done

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,92 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              corresponding to that RtpReceiver in the remote
+              description.</t>
+
+              <t>Construct a table mapping Payload Type to RtpReceiver
```
Payload Type should not be capitalized.

pthatcherg replied:
Done

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,92 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+              <t>Construct a table mapping Payload Type to RtpReceiver
+              for each RtpReceiver configured to receive from this
+              transport and for each Payload Type that RtpReceiver is
+              configured to receive.  The Payload Types of a given
```
This a=fmtp stuff can be omitted; this mapping is discussed in the processing a local description section.

pthatcherg replied:
Done

juberti commented on draft-ietf-rtcweb-jsep.xml:
```diff
@@ -4076,6 +4076,92 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
+                stop.
+              </t>
+              <t>
+                Drop the packet.
```
Perhaps preface with "Otherwise, "

pthatcherg replied:
Done
